### PR TITLE
Move most of Operator<Backend> to OperatorBase. Unify operator interfaces.

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
Operators now have a common interface.
Additionally, `SetupSharedSampleParams` was removed from the common interface and was reinserted in specific use cases after being appropriately renamed.

### Tests:

- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h [hunk 1]
```diff
@@ -358,13 +358,15 @@ class nvJPEGDecoder : public StatelessOperator<MixedBackend>, CachedDecoderImpl
...
   using dali::OperatorBase::Run;
-  void Run(Workspace &ws) override {
+  void RunImpl(Workspace &ws) override {
     SetupSharedSampleParams(ws);
     ParseImagesInfo(ws);
     ProcessImages(ws);
   }
 
  protected:
+  virtual void SetupSharedSampleParams(Workspace &ws) {}
```
Comment (jantonguirao) [inline dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:368]: why is this required?

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h (hunk 1 above)
Comment (mzient) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h:368]: Fused decoders override it.

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/imgcodec/mixed_decoder.cc [hunk 2]
```diff
@@ -23,8 +23,10 @@ class MixedDecoder : public ImageDecoder<MixedBackend> {
   ~MixedDecoder() override = default;
   explicit MixedDecoder(const OpSpec &spec) : ImageDecoder<MixedBackend>(spec) {}
 
-  void Run(Workspace &ws) override {
-    SetupSharedSampleParams(ws);
+  virtual void SetupCropParams(Workspace &ws) {}
```
Comment (stiepan) [inline dali/operators/imgcodec/mixed_decoder.cc:26]: Do we ever overload it?

File: dali/operators/decoder/host/fused/host_decoder_slice.h [hunk 3]
```diff
@@ -29,16 +29,13 @@ class HostDecoderSlice : public HostDecoder {
   inline ~HostDecoderSlice() override = default;
   DISABLE_COPY_MOVE_ASSIGN(HostDecoderSlice);
 
- protected:
-  inline void RunImpl(Workspace &ws) override {
+  using HostDecoder::RunImpl;
```
Comment (stiepan) [inline dali/operators/decoder/host/fused/host_decoder_slice.h:32]: Is this needed?

File: dali/operators/decoder/host/fused/host_decoder_crop.h [hunk 4]
```diff
@@ -29,9 +29,8 @@ class HostDecoderCrop : public HostDecoder, protected CropAttr {
   inline ~HostDecoderCrop() override = default;
   DISABLE_COPY_MOVE_ASSIGN(HostDecoderCrop);
 
-  void inline SetupSharedSampleParams(SampleWorkspace &ws) override {
-    CropAttr::ProcessArguments(spec_, ws);
-  }
+  void RunImpl(Workspace &ws) override;
+  using HostDecoder::RunImpl;
```
Comment (stiepan) [inline dali/operators/decoder/host/fused/host_decoder_crop.h:33]: Is this needed?

File: dali/operators/decoder/host/fused/host_decoder_slice.h (hunk 3 above)
Comment (mzient) [reply dali/operators/decoder/host/fused/host_decoder_slice.h:32]: Yes, otherwise clang is complaining about hiding a function from the superclass.

File: dali/operators/decoder/host/fused/host_decoder_crop.h (hunk 4 above)
Comment (mzient) [reply dali/operators/decoder/host/fused/host_decoder_crop.h:33]: Likewise.

**Review by stiepan — APPROVED** [review]
